compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+     */
+    def expectFailure: CompilationTest =
+      new CompilationTest(targets, times, shouldDelete, threadLimit, true)
+
+    /** Delete all output files generated by this `CompilationTest` */
+    def delete(): Unit = targets.foreach(t => delete(t.outDir))
+
+    private def delete(file: JFile): Unit = {
+      if (file.isDirectory) file.listFiles.foreach(delete)
+      try Files.delete(file.toPath)
+      catch {
+        case _: NoSuchFileException => // already deleted, everything's fine
+      }
+    }
+  }
+
+  /** Create out directory for directory `d` */
+  private def toCompilerDirFromDir(d: JFile, sourceDir: JFile, outDir: String): JFile = {
+    val targetDir = new JFile(outDir + s"${sourceDir.getName}/${d.getName}")
+    targetDir.mkdirs()
+    targetDir
+  }
+
+  /** Create out directory for `file` */
+  private def toCompilerDirFromFile(file: JFile, sourceDir: JFile, outDir: String): JFile = {
+    val uniqueSubdir = file.getName.substring(0, file.getName.lastIndexOf('.'))
+    val targetDir = new JFile(outDir + s"${sourceDir.getName}/$uniqueSubdir")
+    targetDir.mkdirs()
+    targetDir
+  }
+
+  /** Make sure that directory string is as expected */
+  private def checkRequirements(f: String, sourceDir: JFile, outDir: String): Unit = {
+    require(sourceDir.isDirectory && sourceDir.exists, "passed non-directory to `compileFilesInDir`")
+    require(outDir.last == '/', "please specify an `outDir` with a trailing slash")
+  }
+
+  /** Separates directories from files and returns them as `(dirs, files)` */
+  private def compilationTargets(sourceDir: JFile): (List[JFile], List[JFile]) =
+    sourceDir.listFiles.foldLeft((List.empty[JFile], List.empty[JFile])) { case ((dirs, files), f) =>
+      if (f.isDirectory) (f :: dirs, files)
+      else if (f.getName.endsWith(".check")) (dirs, files)
+      else if (f.getName.endsWith(".flags")) (dirs, files)
+      else (dirs, f :: files)
+    }
+
+  /** Gets the name of the calling method via reflection.
+   *
+   *  It does this in a way that needs to work both with the bootstrapped dotty
+   *  and the non-bootstrapped version. Since the two compilers generate
+   *  different bridges, we first need to filter out methods with the same name
+   *  (bridges) - and then find the `@Test` method in our extending class
+   */
+  private def getCallingMethod(): String = {
+    val seen = mutable.Set.empty[String]
+    Thread.currentThread.getStackTrace
+      .filter { elem =>
+        if (seen.contains(elem.getMethodName)) false
+        else { seen += elem.getMethodName; true }
+      }
+      .take(6).find { elem =>
+        val callingClass = Class.forName(elem.getClassName)
+        classOf[ParallelTesting].isAssignableFrom(callingClass) &&
+        elem.getFileName != "ParallelTesting.scala"
+      }
+      .map(_.getMethodName)
+      .getOrElse {
+        throw new IllegalStateException("Unable to reflectively find calling method")
```

Comment:
Most likely it will screw up tests that keep output for compilation that depends on previous compilation - e.g. `tastyBootstrap`.

But yeah - sure, I could make it continue with some randomized value instead.